### DISCUSSION TITLE
how to configure GDL to have more than 100 windows and window positioning is being respected?

### DISCUSSION THREAD

**Discussion by brandy125** [gnudatalanguage/gdl#1378] [Q&A] [answered]
On my aarch64 system I configure fresh GDL like:

build_gdl.sh prep|configure|build

The build shows millions of warnings and notes like:
```
[ 66%] Building CXX object src/CMakeFiles/gdl.dir/libinit_jmg.cpp.o
In file included from /usr/include/GraphicsMagick/Magick++.h:12,
                 from /home/chris/Downloads/gdl-master/src/magick_cl.hpp:21,
                 from /home/chris/Downloads/gdl-master/src/libinit_cl.cpp:37:
/usr/include/GraphicsMagick/Magick++/STL.h:58:17: warning: 'template<class _Arg, class _Result> struct std::unary_function' is deprecated [-Wdeprecated-declarations]
   58 |   : public std::unary_function<Image&,void>
      |                 ^~~~~~~~~~~~~~
/usr/include/GraphicsMagick/Magick++/STL.h:58:17: note: in definition of macro 'MAGICK_UNARY_FUNCTION_IMAGE_REF_BASE'
   58 |   : public std::unary_function<Image&,void>
      |                 ^~~~~~~~~~~~~~
In file included from /usr/include/c++/12/string:48,
                 from /usr/include/c++/12/bits/locale_classes.h:40,
                 from /usr/include/c++/12/bits/ios_base.h:41,
                 from /usr/include/c++/12/ios:42,
                 from /usr/include/c++/12/istream:38,
                 from /usr/include/c++/12/sstream:38,
                 from /usr/include/c++/12/complex:45,
                 from /usr/include/eigen3/Eigen/Core:50,
                 from /home/chris/Downloads/gdl-master/src/includefirst.hpp:50,
                 from /home/chris/Downloads/gdl-master/src/libinit_cl.cpp:26:
/usr/include/c++/12/bits/stl_function.h:117:12: note: declared here
  117 |     struct unary_function
      |            ^~~~~~~~~~~~~~
/usr/include/GraphicsMagick/Magick++/STL.h:58:17: warning: 'template<class _Arg, class _Result> struct std::unary_function' is deprecated [-Wdeprecated-declarations]
   58 |   : public std::unary_function<Image&,void>
      |                 ^~~~~~~~~~~~~~
/usr/include/GraphicsMagick/Magick++/STL.h:58:17: note: in definition of macro 'MAGICK_UNARY_FUNCTION_IMAGE_REF_BASE'
   58 |   : public std::unary_function<Image&,void>
      |                 ^~~~~~~~~~~~~~
/usr/include/c++/12/bits/stl_function.h:117:12: note: declared here
  117 |     struct unary_function
      |            ^~~~~~~~~~~~~~
/usr/include/GraphicsMagick/Magick++/STL.h:58:17: warning: 'template<class _Arg, class _Result> struct std::unary_function' is deprecated [-Wdeprecated-declarations]
   58 |   : public std::unary_function<Image&,void>
      |                 ^~~~~~~~~~~~~~
/usr/include/GraphicsMagick/Magick++/STL.h:58:17: note: in definition of macro 'MAGICK_UNARY_FUNCTION_IMAGE_REF_BASE'
   58 |   : public std::unary_function<Image&,void>
      |                 ^~~~~~~~~~~~~~
```

When I run GDL now I can create only 100 windows:
```
GDL> for i=0,200 do begin print,i& window,/free & endfor
...
      68
      69
% WINDOW: No more window handles left.
% Execution halted at: $MAIN$          
GDL> 
```
This was the first problem.
When I want to place a window in a specific position then it is not being respected:
```
GDL> window,1,xpos=0,ypos=0,/free
```
positions the window in the middle of the screen.

When I compile with

```
cmake .. -DX11=ON
```
and use 

```
gdl --no-use-wx
```
Then I am still limited to 100 windows but the window positioning is working.
Beside this the maximum window size is the size of the screen, independend from the given size:

```
GDL>window,1,xsize=5000,ysize=5000
```
opens just a window in the size of my screen (3840x2160)

I am a little lost now. 
What is the correct procedure to get:
a) more than 100 windows
b) window positioning is being respected
c) window size is begin respected

Thanks for any help in advance

**Comment by GillesDuvert** [comment, marked as answer]
@brandy125 the max number of windows IS a problem and a difficult one to tackle in GDL. Sorry for that.

Concerning the size and positions:

GDL has a preferred 'unified' graphics interface, wxWidgets, that makes both plot windows and widget (graphic) windows.
With option --no-use-wx (and if compiled with -DX11) it uses X11, not wxWidgets, for plot windows. X11 is more basic than wxWidgets and the commands go through an entirely different code to make plots. The **X11 code does not permit to make windows wider than the screen**, wxWidgets **does** (using scrollbars).
X11 is very basic and goes over the head of the window manager to create plot windows of any size (less that screen!) at any position.
wxWidgets follows the opendesktop guidelines (or whatever) and its size and position orders MAY be countered by the window manager (focus stealing etc...). So with wxWidgets, it may be useful to tell your favourite window manager to let GDL be.

On my machine, both wxWidgets and X11 make windows of the requested size at the requested position, so it may be necessary to check the window manager preferences to see if something prevent the placement of windows.

And, graphicsMagick use deprecated C++ functions, wonder why C++ deprecate things and annoy programmers.

**Comment by GillesDuvert** [comment]
I should add that the max number of windows is a plplot limitation, not a GDL one.

**Comment by brandy125** [comment]
@GillesDuvert Thanks for answering. If the number of windows is a plplot limitation, could it be increased then by modifying the code?

Other question, the following plot is very very slow on my machine. Do you have any hint what I can do to improve speed?
```
line=fltarr(1000000) & tic & plot,line,yrange=[-1,1],psym=1 & toc
```
IDL on Suse inside Mac Mini with Intel chip: 0.4 sec
GDL on Suse inside Mac Mini with Intel chip: 5 sec
GDL on Suse inside Mac Studio with Ultra (M1) chip and wxwidget: 34 sec
GDL on Suse inside Mac Studio with Ultra (M1) chip and X11: 78 sec
 
Yes, I have plots with more than 10.000.000 points that would need 10 mins to finish.

I know it is hard to say what is the problem but maybe you have ideas.


Just for completenes:
FL    on Suse inside Mac Studio with Ultra (M1) chip and X11: 0.4 sec
```
bash$ fl
Fawlty Language 0.79.52 (linux aarch64 m64) Copyright (c) 2000-2022, Lajos Foldy
                         https://www.flxpert.hu/fl/
*** THIS SOFTWARE COMES WITH ABSOLUTELY NO WARRANTY! USE AT YOUR OWN RISK! ***

Type 'help, /lib' for system routines info.

FL> line=fltarr(1000000) & tic & plot,line,yrange=[-1,1],psym=1 & toc
libGL error: pci id for fd 7: 1ab8:0010, driver (null)
pci id for fd 8: 1ab8:0010, driver (null)
% Time elapsed: 0.453301 seconds.
FL> 
```

Why can Fawlty Language do it so fast? This gives me hope that there is something that I could do.

**Reply by alaingdl** [reply]
We know that since a very long time, when we have psym symbols. Plots without psym are fast. Yes it is s serious bottleneck.

I don't know if it is a problem of Plplot or a bad way to call plplot from internal GDL code. (Maybe some internal loop in GDL we may pass directly to plplot ?!)


If we establish a list of urgent tasks & corrections, for sure this problem would be in the list, somewhere on the top

**Reply by GillesDuvert** [reply]
I do not know how FL or IDL do the plots, but this is no surprise: PSYMs in GDL are internally interpreted as small drawings in **user coordinates**. Meaning that for a cross there is a computation of 4 or more points to be drawn for each of the 1000000 elements, and as many calls to the plotting function --- compared to 1 call and nothing else for a PSYM=0 (line). 
But, the plotting function is a complicated loop containg a number of if() inside.
Removing these internal if() will certainly speedup plots a lot.

**Reply by GillesDuvert** [reply]
#1385 is proposed, normally it should speedup plots.
